ethapi/config.go

```diff
@@ -0,0 +1,138 @@
+	BlobSchedule *params.BlobConfig `json:"blobSchedule"`
+
+	ChainId *hexutil.Big `json:"chainId"`
+
+	// ForkId in sonic is a checksum derived from the json marshall of the corresponding upgrade
+	ForkId hexutil.Bytes `json:"forkId"`
+
+	Precompiles     contractRegistry `json:"precompiles"`
+	SystemContracts contractRegistry `json:"systemContracts"`
+}
+
+// helper types to improve readability of the returned structure.
+type contractRegistry map[string]common.Address
+type forkId [4]byte
+
+// configResponse is the response structure for the Config method as
+// described by https://eips.ethereum.org/EIPS/eip-7910
+type configResponse struct {
+	// current is the config active at the latest block number.
+	Current *config `json:"current"`
+	// Next will remain nil since Sonic config activation does not depend on time.
+	Next *config `json:"next"`
+	// Last could be nill if only one upgrades heights exists.
+	Last *config `json:"last"`
+}
+
+// makeConfigFromUpgrade constructs the config that was active for the
+// given block number based on the upgrade heights.
+func makeConfigFromUpgrade(
+	ctx context.Context,
+	b Backend,
+	upgradeHeight opera.UpgradeHeight,
+) (*config, error) {
+
+	chainID := b.ChainID()
+	chainCfg := opera.CreateTransientEvmChainConfig(
+		chainID.Uint64(),
+		[]opera.UpgradeHeight{upgradeHeight},
+		upgradeHeight.Height,
+	)
+
+	precompiled := make(contractRegistry)
+	chainCfgRules := chainCfg.Rules(big.NewInt(int64(upgradeHeight.Height)), true, uint64(0))
+	for addr, c := range vm.ActivePrecompiledContracts(chainCfgRules) {
+		precompiled[c.Name()] = addr
+	}
+
+	forkId, err := makeForkId(upgradeHeight.Upgrades)
+	if err != nil {
+		// this can only fail if json.Marshal fails, which is unexpected
+		return nil, fmt.Errorf("could not make fork id, %v", err)
+	}
+
+	block, err := b.BlockByNumber(ctx, (rpc.BlockNumber(int64(upgradeHeight.Height))))
+	if err != nil {
+		return nil, fmt.Errorf("could not get block %d to determine activation time, %v", upgradeHeight.Height, err)
+	}
+
+	return &config{
+		// block time needs to be converted to unix timestamp as it is done in
+		// evmcore/dummy_block.go in method EvmHeader.EthHeader()
+		ActivationTime:  uint64(block.Time.Unix()),
+		ChainId:         (*hexutil.Big)(chainID),
+		ForkId:          forkId[:],
+		Precompiles:     precompiled,
+		SystemContracts: activeSystemContracts(upgradeHeight.Upgrades),
+	}, nil
+}
+
+// activeSystemContracts returns a map of system contract names to their addresses
+// based on the active upgrade.
+func activeSystemContracts(upgrade opera.Upgrades) contractRegistry {
+	sysContracts := make(contractRegistry)
+	if upgrade.Allegro {
+		sysContracts["HISTORY_STORAGE"] = params.HistoryStorageAddress
+	}
+	if upgrade.GasSubsidies {
+		sysContracts["GAS_SUBSIDY_REGISTRY"] = registry.GetAddress()
+	}
+	return sysContracts
+}
+
+// makeForkId creates a fork ID from the given upgrade.
+func makeForkId(upgrade opera.Upgrades) (forkId, error) {
```

Comment:
According to [EIP-6122](https://eips.ethereum.org/EIPS/eip-6122#specification) the `ForkId` should also contain the genesis hash of the chain, presumably to differentiate similar forks on different chains (e.g. on the sonic main-net and various test nets). I think we should include this. Wdyt?